src/DryIoc/Container.cs

```diff
@@ -9561,22 +9652,63 @@ public virtual Expression GetExpressionOrDefault(Request request)
 
                     serviceExpr = ApplyReuse(serviceExpr, request);
 
-                    if (serviceExpr.Type != originalServiceExprType)
+                    if (serviceExpr.NodeType != ExprType.Constant &&
+                        serviceExpr.Type != originalServiceExprType && 
+                        !originalServiceExprType.GetTypeInfo().IsAssignableFrom(serviceExpr.Type.GetTypeInfo()))
                         serviceExpr = Convert(serviceExpr, originalServiceExprType);
                 }
+                else
+                {
+                    // it does not make sense to split the scoped or singleton graphs
+                    // because former already includes the lambda and latter is a constant or lambda
+                    if (reuse == DryIoc.Reuse.Transient &&
+                        !rules.UsedForValidation && !rules.UsedForExpressionGeneration) 
+                    {
+                        // Split the expression with dependencies bigger than certain threshold by wrapping it in Func which is a
+                        // separate compilation unit and invoking it emmediately
+                        var depCount = request.DependencyCount;
+                        if (depCount >= 1024)
```

Comment:
@dadhi What is this number 1024 based on? Is there any reason why its better than 128? or any other number? This to me looks some kind of treshold variable, so should this be configured so we can test it looping through different values?